src/main/java/net/minecraftforge/client/SkyRenderer.java

```diff
@@ -0,0 +1,490 @@
+import net.minecraft.client.renderer.BufferBuilder;
+import net.minecraft.client.renderer.GlStateManager;
+import net.minecraft.client.renderer.RenderHelper;
+import net.minecraft.client.renderer.Tessellator;
+import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
+import net.minecraft.client.renderer.vertex.VertexBuffer;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
+
+public class SkyRenderer
+{
+    private EnumMap<EnumSkyPass, PartSkyRenderer> renderers = new EnumMap<EnumSkyPass, PartSkyRenderer>(EnumSkyPass.class)
+    {{
+        this.put(EnumSkyPass.SKY_BACK, new PartSkyRenderer(getSkyBackDefault()));
+        this.put(EnumSkyPass.SUN_MOON_STARS, new PartSkyRenderer(getSunMoonStarsDefault()));
+        this.put(EnumSkyPass.SKY_FRONT, new PartSkyRenderer(getSkyFrontDefault()));
+    }};
+
+    public void setSkyRenderer(EnumSkyPass pass, ResourceLocation id, IRenderHandler renderer, double priority)
+    {
+        renderers.get(pass).setSkyRenderer(id, renderer, priority);
+    }
+
+    public boolean render(float partialTicks, WorldClient world, Minecraft mc)
+    {
+        boolean enabled = false;
+        for(PartSkyRenderer part : renderers.values())
+            enabled = enabled || part.enabled();
+
+        if(!enabled)
+            return false;
+
+        renderers.get(EnumSkyPass.SKY_BACK).render(partialTicks, world, mc);
+        renderers.get(EnumSkyPass.SUN_MOON_STARS).render(partialTicks, world, mc);
+        renderers.get(EnumSkyPass.SKY_FRONT).render(partialTicks, world, mc);
+        return true;
+    }
+
+    private static class PartSkyRenderer {
+        Map<ResourceLocation, RenderHandler> defHandlers;
+        SortedMap<ResourceLocation, RenderHandler> handlers = Maps.newTreeMap();
+
+        PartSkyRenderer(Map<ResourceLocation, RenderHandler> defaultHandlers) {
+            this.defHandlers = defaultHandlers;
+        }
+
+        void setSkyRenderer(ResourceLocation id, IRenderHandler renderer, double priority)
+        {
+            if(renderer != null)
+                handlers.put(id, new RenderHandler(renderer, priority));
+            else if(defHandlers.containsKey(id))
+                handlers.put(id, defHandlers.get(id));
+            else handlers.remove(id);
+        }
+
+        boolean enabled() {
+            return !handlers.equals(this.defHandlers);
+        }
+
+        void render(float partialTicks, WorldClient world, Minecraft mc)
+        {
+            for(RenderHandler handler : handlers.values())
+            {
+                handler.handler.render(partialTicks, world, mc);
+            }
+        }
+    }
+
+    private static class RenderHandler implements Comparable<RenderHandler> {
+        IRenderHandler handler;
+        double priority;
+
+        RenderHandler(IRenderHandler handler, double priority)
+        {
+            this.handler = handler;
+            this.priority = priority;
```

Comment:
Could priority be simplified to an enum of priority types, similar to how event priorities are handled?

Reply:
No. It's priority like that applied on IWorldGenerator. It can't be replaced with enums because modders will want to add their own ordered renderers with arbitrary size.

Reply:
👍